### ISSUE TITLE
HTCondor Batch System: Add fileStore support for pools without shared fs (e.g. OSG)

### ISSUE THREAD

**Issue by jasoncpatton** [BD2KGenomics/toil#2132]
The current implementation of the HTCondor batch system in Toil, when using a fileStore (file jobStore), works with HTCondor clusters (or "pools") that have a shared filesystem. As with other batch systems, the Toil library (e.g. via a user's virtualenv) and user's fileStore must have the same absolute paths across execute nodes.

HTCondor is different from many other batch systems in that shared file systems are not a requirement when building a HTCondor pool. When submitting jobs to a HTCondor pool from a submit node, users specify their executable and any input files (and, optionally, output files). By default, if the execute node that the job lands on does not share a filesystem with the submit node, the executable and input files are transferred to the execute node in a sandbox directory, the job is executed from the sandbox, and output files are transferred back to the submit node when the job completes. The no-shared-fs scenario is the case for one of the most widely accessible HTCondor pools -- the Open Science Grid (OSG).

Three problems must be solved to add support for using the HTCondor batch system with fileStores when submitting jobs via Toil:

1. **Determining input files.** Currently, a job node may create any files (`fileStoreID`s) and access any files (via `fileStoreID`s) in a fileStore, i.e. input and output files are not a property of a single job node (except, perhaps, any temp files created). A fileStore does not account for which job nodes add/access/remove files to/from it. If a job node (e.g. a wrapped function) needs to know about certain files, `fileStoreID`s are typically passed to it via arbitrary arguments.
1. **Rewriting absolute paths.** A fileStore is represented by an absolute path. This absolute path is passed in arguments to `_toil_worker` via `JobNode.command` and to `_toil` via `JobGraph.command`. When executing `_toil_worker` and/or `_toil` on an execute node, the fileStore will instead be located in the sandbox (`$_CONDOR_SCRATCH_DIR/MyFileStore`) so the helper scripts will need to be pointed there instead.
1. **Deploying Toil on an execute node.** Before a job node can run on an execute node, the Toil library (including `_toil_worker` and `_toil`) must be installed and in the path of the job environment.

For (1), without some major changes to the Toil code, it seems like we are stuck with transferring the entire fileStore to/from the execute node. There is simply no way to guarantee from the objects passed to the batch system (`jobNode`, `config`, etc.) that a job node will have all of the input files it needs from the fileStore. I would greatly appreciate if someone has any ideas about a better option here. Another jobStore class specifically for HTCondor jobs?

(2) and (3) may be able to be done with a pre-script, which would work similar to Toil's hot/remote deployment. An HTCondor job submission can be given a script to run before the job executes but after files have been transferred. Users would have to supply a Python tarball with Toil (or virtualenv to create an environment to `pip install toil` in), which would be unpacked and have its path(s) added to the job's environment. This script could also unpickle/edit/repickle the `JobNode`/`JobGraph` objects.

**Comment by DailyDreaming** [comment]
For point (1), I'd like to see files tracked as inputs/outputs as a part of the fileStore proper.  This need not be HTCondor specific.  This would take a bit of work I imagine, but has utility in other areas, particularly debugging.

For (2) and (3) in general, I like the hot-deployment idea, though our current implementation generally expects to find itself within a docker image within a folder with the same name, even if the filesystem has changed.  That might be a path forward, but I like your idea of checking the filesystem and rewriting the absolute paths as needed.  We could integrate scripts into Toil proper as functions specifically to be run for HTCondor at different stages.  I'd like other opinions on this as well though.

**Comment by jasoncpatton** [comment]
Funny enough, we have a "docker universe" in HTCondor which lets users run their jobs within a specified image, but it requires Docker be installed on all the execute nodes and that the pool admin enable docker universe, which is not the case right now in OSG. (Maybe in a few months??? I'd have to ask the OSG people.) But we'd still have an issue because HTCondor bind mounts the sandbox directory on the host machine to its exact same path (usually something like `/var/lib/condor/execute/dir_12345`) within a container, and a job runs as a user within a container, not as root, so it would difficult inside a job to rebind as needed (e.g. to bind `/home/jasoncpatton/toil/MyJobStore` to `$_CONDOR_SCRATCH_DIR/MyJobStore`).

**Comment by DailyDreaming** [comment]
Yes, different bind mounts would be a problem that we'd have to write for.  Toil's developed and run with the expectation that the jobStore not move.  In terms of how we scale with different file systems on the cloud, we usually designate one s3/google bucket and this filesystem hosts the jobStore and all worker nodes leave their outputs here and coordinate with the leader through this permanent jobStore indirectly.  Workers spin up and down leaving their outputs in this jobStore, which can become inputs for further jobs spinning up on instances later.  I'm not sure about implementing Toil in this way on HTCondor and this would have to basically coordinate a new provisioner system that would coordinate with HTCondor I imagine, and I think the scripts you've suggested are currently the best way forward.  Again though, I'd like to hear other ideas on this as well.

**Comment by mr-c** [comment]
For
> 1. **Determining input files.** Currently, a job node may create any files (fileStoreIDs) and access any files (via fileStoreIDs) in a fileStore, i.e. input and output files are not a property of a single job node (except, perhaps, any temp files created). A fileStore does not account for which job nodes add/access/remove files to/from it. If a job node (e.g. a wrapped function) needs to know about certain files, fileStoreIDs are typically passed to it via arbitrary arguments.

in CWL mode (`toil-cwl-runner`) we do know exactly which files are needed for a particular job

**Comment by jasoncpatton** [comment]
For a new jobStore type, HTCondor does support the [Chirp protocol](http://research.cs.wisc.edu/htcondor/manual/current/condor_chirp.html) for transferring files to/from the submit (leader) node, e.g.:
```
condor_chirp fetch /home/jcpatton/myJobStore/tmp/x/X/tmpABCD ./myJobStore/tmp/x/X/tmpABCD
condor_chirp put ./myJobStore/tmp/y/Y/tmpEFGH /home/jcpatton/myJobStore/tmp/y/Y/tmpEFGH
```

However, from our perspective, this is _not_ an ideal solution because condor_chirp file transfers are not scheduled and because transferring large files using condor_chirp is not recommended. I might still take a shot at implementing a "chirpStore" as a first try at an on-demand file transfer solution.

> in CWL mode (`toil-cwl-runner`) we do know exactly which files are needed for a particular job

It's not clear to me how to access that information from inside a `HTCondorBatchSystem` object, other than to parse `inputs:` from the CWL file itself (which we can find from the jobNode). It appears that the files are first staged into a shared jobStore regardless?

**Comment by mr-c** [comment]
> > in CWL mode (`toil-cwl-runner`) we do know exactly which files are needed for a particular job

> It's not clear to me how to access that information from inside a `HTCondorBatchSystem` object, other than to parse `inputs:` from the CWL file itself (which we can find from the jobNode). It appears that the files are first staged into a shared jobStore regardless?

I'm not a core Toil dev, so I can't quote the API, but if the information that CWL has isn't exposed conveniently then we could work together to provide that. If CWL is motivating the use of Toil with HTCondor then I suggest a conference call between the tool devs, myself, and you.

**Comment by jasoncpatton** [comment]
> I'm not a core Toil dev, so I can't quote the API, but if the information that CWL has isn't exposed conveniently then we could work together to provide that. If CWL is motivating the use of Toil with HTCondor then I suggest a conference call between the tool devs, myself, and you.

CWL support is definitely a motivating factor for adding HTCondor support. I've added my email to my profile, otherwise @ejacox also has our contact info.

**Comment by jasoncpatton** [comment]
Todo items from today's discussion:

1. Add support for scheduling local jobs.
1. Explore the CWLJob object and CWL intermediate jobs.
1. Look at output to cleanup fileStore before transferring output files.

**Comment by adamnovak** [comment]
Do we still want to develop a dedicated HTCondor-optimized job store that tries to use HTCondor's notion of input and output files to ship files around?
